pkg/shared/utils/containerRootfs/fs.go

```diff
@@ -0,0 +1,166 @@
+package containerRootfs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/anchore/stereoscope"
+	"github.com/anchore/stereoscope/pkg/file"
+	"github.com/anchore/stereoscope/pkg/filetree/filenode"
+)
+
+const perFileReadLimit = 2 * file.GB
+
+func ToDirectory(ctx context.Context, src, dest string) error {
+	image, err := stereoscope.GetImage(ctx, src)
+	if err != nil {
+		return fmt.Errorf("unable to parse image from src %s: %w", src, err)
+	}
+	defer image.Cleanup()
+
+	err = image.SquashedTree().Walk(func(path file.Path, f filenode.FileNode) error {
+		target := filepath.Join(dest, string(path))
+
+		switch f.FileType {
+		case file.TypeDirectory:
+			if err := os.MkdirAll(target, 0755); err != nil {
+				return err
+			}
+		case file.TypeRegular:
+			indexEntry, err := image.FileCatalog.Get(*f.Reference)
+			if err != nil {
+				return fmt.Errorf("unable to get index entry for file: %w", err)
+			}
+
+			output, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, indexEntry.Mode())
+			if err != nil {
+				return fmt.Errorf("unable to open file %s: %w", target, err)
+			}
+
+			content, err := image.OpenReference(*f.Reference)
+			if err != nil {
+				return fmt.Errorf("unable to open path from squash: %w", err)
+			}
+
+			numBytes, err := io.Copy(output, io.LimitReader(content, perFileReadLimit))
+			if numBytes >= perFileReadLimit || errors.Is(err, io.EOF) {
+				return fmt.Errorf("zip read limit hit (potential decompression bomb attack)")
+			}
+			if err != nil {
+				return fmt.Errorf("unable to copy file: %w", err)
+			}
+
+			err = output.Close()
+			if err != nil {
+				// TODO log the error
+			}
+		}
+
+		return nil
+	}, nil)
+	if err != nil {
+		return fmt.Errorf("unable to walk squashed tree from container: %w", err)
+	}
+
+	return nil
+}
+
+type Rootfs interface {
+	Dir() string
+	Cleanup() error
+}
+
+func ToTempDirectory(ctx context.Context, src string) (Rootfs, error) {
+	tdrf := &TempDirRootfs{}
+
+	tmpdir, err := os.MkdirTemp("", "")
+	if err != nil {
+		return tdrf, fmt.Errorf("unable to create temp directory: %w", err)
+	}
+	tdrf.dir = tmpdir
+
+	err = ToDirectory(ctx, src, tmpdir)
+	if err != nil {
+		return tdrf, fmt.Errorf("unable output squashed image to directory: %w", err)
+	}
+
+	return tdrf, nil
+}
+
+type TempDirRootfs struct {
+	dir string
+}
+
+func (tdrf *TempDirRootfs) Dir() string {
+	return tdrf.dir
+}
+
+func (tdrf *TempDirRootfs) Cleanup() error {
+	var errs []error
+
+	if tdrf.dir != "" {
+		err := os.RemoveAll(tdrf.dir)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("unable to remove temp directory: %w", err))
+		}
+	}
+
+	return errors.Join(errs...)
```

Comment:
why we need `errs` and `Join` if we have only single error here?

Reply:
Originally I had multiple things getting cleaned up here, but now its just the one thing. I'll refactor this.